dandiapi/api/doi.py

```diff
@@ -0,0 +1,63 @@
+from datetime import datetime
+from typing import Dict
+
+from django.conf import settings
+import requests
+
+from dandiapi.api.models import Version
+
+
+def _generate_doi_data(version: Version) -> Dict:
+    prefix = settings.DANDI_DOI_API_PREFIX
+    dandiset_id = version.dandiset.identifier
+    version_id = version.version
+    doi = f'{prefix}/{dandiset_id}/{version_id}'
+    url = f'https://dandiarchive.org/dandiset/{dandiset_id}/{version_id}'
+    creators = []
+    contributors = []
+    if 'contributor' in version.metadata.metadata:
+        creators = [
+            {'name': contributor['name']}
+            for contributor in version.metadata.metadata['contributor']
+            if 'dandi:Author' in contributor['roleName']
+        ]
+        contributors = [
+            {'name': contributor['name']}
+            for contributor in version.metadata.metadata['contributor']
+        ]
+    return {
+        'data': {
+            'id': doi,
+            'type': 'dois',
+            'attributes': {
+                'doi': doi,
+                'creators': creators,
+                'titles': [{'title': version.name}],
+                'publisher': 'DANDI Archive',
+                'publicationYear': datetime.now().year,
+                'contributors': contributors,
+                'types': {'resourceTypeGeneral': 'NWB'},
+                'url': url,
+            },
+        }
+    }
+
+
+def create_doi(version: Version):
+    # If DOI isn't configured, skip this step
+    if (
+        settings.DANDI_DOI_API_URL is None
+        and settings.DANDI_DOI_API_USER is None
+        and settings.DANDI_DOI_API_PASSWORD is None
+        and settings.DANDI_DOI_API_PREFIX is None
+    ):
+        return
+    request_body = _generate_doi_data(version)
+    requests.post(
+        settings.DANDI_DOI_API_URL,
+        json=request_body,
+        auth=requests.auth.HTTPBasicAuth(
+            settings.DANDI_DOI_API_USER,
+            settings.DANDI_DOI_API_PASSWORD,
+        ),
+    ).raise_for_status()
```

Comment:
I guess (didn't check) that as a result of that post you should receive an actual DOI and that is the one then which should be assigned to the `doi` field within metadata of the published dandiset.

Also relating to above comment on `f'{prefix}/{dandiset_id}/{version_id}'` - I did not know that it is legit for suffix (everything after `{prefix}/`) to contain `/` .  I do see though that there is now DOI 10.80507/000001/0.210323.2052 on that test sandbox instance and according to https://www.doi.org/doi_handbook/2_Numbering.html

2.6.3 URN presentation

To enable the use of DOIs in workflows that have already standardized on URNs, the DOI proxy servers understand the substitution of a colon in place of the initial slash in a DOI name. DOI names may therefore be expressed as URNs in the doi.org domain by writing, for example, the DOI name 10.123/456 in the form https://doi.org/urn:doi:10.123:456. Note, however, that a DOI suffix is allowed to contain other slashes, and where these occur they must be percent-encoded rather than replaced with a colon: for example, the DOI name 10.123/456ABC/zyz would become https://doi.org/urn:doi:10.123:456ABC%2Fzyz, with the final slash character encoded as %2F.

So may be this "complication" is why typically people avoid using `/` in the DOI suffixes?  E.g. openneuro has `10.18112/openneuro.ds003569.v1.0.0` for https://openneuro.org/datasets/ds003569/versions/1.0.0

@satra -- are you aware of gotchas with `/` in the DOI suffix and why it seems generally is avoided?

Reply:
I'm happy to change this to whatever. I thought `10.80507/000001/0.210323.2052` would match how the `dandiset.yaml` files are stored.

Reply:
I also like how it looks, but as I have mentioned, I did not even know that it is possible since I never saw any DOI with `/` in the suffix.  So we need to research on that more.  @effigies may be you know or can refer us to someone who would be able to answer the question  
"why (is there any pragmatic reason) openneuro has DOI like 10.18112/openneuro.ds003569.v1.0.0 and not e.g. 10.18112/openneuro/ds003569/v1.0.0"?

Reply:
Regarding openneuro DOIs, when the code to mint DOIs was first written I didn't know that '/' could be used in the suffix and for whatever reason I didn't look it up. '.' as a delimiter was common in the Datacite documentation so I defaulted to that.

Reply:
Thank you @rwblair!